edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/service/PortExtensions.cs

```diff
@@ -2,56 +2,50 @@
-                {
-                    if (int.TryParse(portProtocol[0], out int port) && TryValidateProtocol(portProtocol[1], out string protocol))
-                    {
-                        serviceList.Add((port, protocol));
-                    }
-                    else
-                    {
-                        Events.InvalidExposedPortValue(exposedPort.Key);
-                    }
-                }
-            }
+        public static List<V1ContainerPort> GetContainerPorts(this IDictionary<string, global::Docker.DotNet.Models.EmptyStruct> ports)
+            => ports.Select(port => ExtractContainerPort(port.Key)).FilterMap().ToList();
 
-            return serviceList.Count > 0 ? Option.Some(serviceList) : Option.None<List<(int, string)>>();
-        }
+        static Option<V1ContainerPort> ExtractContainerPort(string exposedPort)
+            => PortAndProtocol.Parse(exposedPort)
+                .Map(portAndProtocol => new V1ContainerPort(portAndProtocol.Port, protocol: portAndProtocol.Protocol));
 
-        public static bool TryValidateProtocol(string dockerProtocol, out string k8SProtocol)
-        {
-            bool result = true;
-            switch (dockerProtocol.ToUpper())
-            {
-                case "TCP":
-                    k8SProtocol = "TCP";
-                    break;
-                case "UDP":
-                    k8SProtocol = "UDP";
-                    break;
-                case "SCTP":
-                    k8SProtocol = "SCTP";
-                    break;
-                default:
-                    k8SProtocol = "TCP";
-                    result = false;
-                    break;
-            }
+        public static List<V1ServicePort> GetExposedPorts(this IDictionary<string, global::Docker.DotNet.Models.EmptyStruct> ports)
+            => ports.Select(port => ExtractExposedPort(port.Key)).FilterMap().ToList();
 
-            return result;
-        }
+        static Option<V1ServicePort> ExtractExposedPort(string exposedPort) =>
+            PortAndProtocol.Parse(exposedPort)
+                .Map(portAndProtocol => new V1ServicePort(portAndProtocol.Port, name: $"ExposedPort-{portAndProtocol.Port}-{portAndProtocol.Protocol}".ToLowerInvariant(), protocol: portAndProtocol.Protocol));
+
+        public static List<V1ServicePort> GetHostPorts(this IDictionary<string, IList<PortBinding>> ports)
+            => ports.SelectMany(port => ExtractHostPorts(port.Key, port.Value)).ToList();
+
+        static IEnumerable<V1ServicePort> ExtractHostPorts(string name, IList<PortBinding> bindings)
+            =>
+                PortAndProtocol.Parse(name)
+                    .Map(
+                        portAndProtocol =>
+                            bindings.Select(
+                                    hostBinding =>
+                                    {
+                                        if (int.TryParse(hostBinding.HostPort, out int hostPort))
```

Comment:
>HostPort [](start = 69, length = 8)

Apparently the TryValidateProtocol was implicitly defaulting to "TCP. Does this behavior still hold true? If yes, add a UT for this behavior?

Reply:
Behavior is not changed. It was just a bit confusing before. It use to be `bool TryValidateProtocol(key, out protocol)` method that returned false if user specified unsupported protocol. In this case TCP was set to TCP, but return value is `false`. 
Code that used TryValidateProtocol checked return value and didn't apply protocol value.

Now intention declare more explicitly if protocol is not supported it returns `None`